### ISSUE TITLE
URI error while decoding base64 HTML

### ISSUE THREAD

**Issue by Sephaq** [dankogai/js-base64#130]
**Version:** `3.4.5`
### **Problem**

The following Base64 HTML is giving me a crash when I try to decode it using **Base64.decode(string)** function:
```
PHAgc3R5bGU9InRleHQtYWxpZ246bGVmdDsiPjxzcGFuIHN0eWxlPSJjb2xvcjogcmdiKDE3LDE3LDE3KTtiYWNrZ3JvdW5kLWNvbG9yOiByZ2IoMjU1LDI1NSwyNTUpO2ZvbnQtc2l6ZTogMTNweDsiPldoYXRldmVyIHRoZSBzZWFzb24sIHBvd2VyIG91dGFnZXMgYXJlIGEgZmFjdCBvZiBsaWZlLCB3aGljaCBpcyB3aHkgaGF2aW5nIGEgYmFja3VwIGdlbmVyYXRvciBpcyBhIGdyZWF0IHdheSB0byBrZWVwIHlvdSBwb3dlcmVkIHVwIGFuZCB5b3VyIGFwcGxpYW5jZXMgcnVubmluZyBubyBtYXR0ZXIgd2hhdCBoYXBwZW5zLiBQaWNrIG91dCB0aGUgcmlnaHQgZ2VuZXJhdG9yIGZvciB5b3UgYnkga2VlcGluZyB0aGVzZSBoZWxwZnVsIGZhY3RzIGluIG1pbmQuPC9zcGFuPjwvcD4KPHVsPgo8bGkgc3R5bGU9InRleHQtYWxpZ246bGVmdDsiPjxzcGFuIHN0eWxlPSJjb2xvcjogcmdiKDE3LDE3LDE3KTtiYWNrZ3JvdW5kLWNvbG9yOiByZ2IoMjU1LDI1NSwyNTUpO2ZvbnQtc2l6ZTogMTNweDsiPjxzdHJvbmc+R2VuZXJhdG9yIHR5cGVzLjwvc3Ryb25nPiBUaGVyZSBhcmUgdHdvIHR5cGVzIG9mIGdlbmVyYXRvcnMgY29udmVudGlvbmFsIGFuZCBpbnZlcnRlci48L3NwYW4+PC9saT4KPC91bD4KPG9sPgo8bGkgc3R5bGU9InRleHQtYWxpZ246bGVmdDsiPjxzcGFuIHN0eWxlPSJjb2xvcjogcmdiKDE3LDE3LDE3KTtiYWNrZ3JvdW5kLWNvbG9yOiByZ2IoMjU1LDI1NSwyNTUpO2ZvbnQtc2l6ZTogMTNweDsiPkEgPHN0cm9uZz5jb252ZW50aW9uYWwgZ2VuZXJhdG9yPC9zdHJvbmc+IGlzIHR5cGljYWxseSBsb3VkZXIgYW5kIGxhcmdlciBhbmQgcnVucyBvbiBnYXNvbGluZS4gVGhleSBhcmUgdXN1YWxseSBnb29kIGF0IHBvd2VyaW5nIGEgcG9ydGlvbiBvZiB5b3VyIGhvbWUgKHVwIHRvIDMgb3IgNCByb29tcykuIENvbnZlbnRpb25hbCBnZW5lcmF0b3JzIGFyZSBncmVhdCBmb3IgcmVmcmlnZXJhdG9ycywgZWxlY3RyaWMgc3RvdmUgdG9wcywgQUMgdW5pdHMgYW5kIG90aGVyIHF1YWxpdHkgb2YgbGlmZSBkZXZpY2VzLjwvc3Bhbj48L2xpPgo8bGkgc3R5bGU9InRleHQtYWxpZ246bGVmdDsiPjxzcGFuIHN0eWxlPSJjb2xvcjogcmdiKDE3LDE3LDE3KTtiYWNrZ3JvdW5kLWNvbG9yOiByZ2IoMjU1LDI1NSwyNTUpO2ZvbnQtc2l6ZTogMTNweDsiPkFuIDxzdHJvbmc+aW52ZXJ0ZXIgZ2VuZXJhdG9yPC9zdHJvbmc+IGlzIHNtYWxsZXIgYW5kIGNhbiBydW4gb24gZ2Fzb2xpbmUgb3IgcHJvcGFuZS4gVGhlc2UgYXJlIHVzdWFsbHkgZ29vZCBmb3IgcG93ZXJpbmcgb25lIG9yIHR3byBkZXZpY2VzLiBBbiBpbnZlcnRlciBnZW5lcmF0b3IgaXMgaWRlYWwgZm9yIHBvd2VyaW5nIGNvbXB1dGVycywgY2VsbHBob25lcywgcmFkaW9zLCBhbmQgb3RoZXIgZGV2aWNlcyB0aGF0IGFyZSBjcml0aWNhbCBkdXJpbmcgZW1lcmdlbmNpZXMuPC9zcGFuPjwvbGk+Cjwvb2w+Cjx1bD4KPGxpPjxzcGFuIHN0eWxlPSJjb2xvcjogcmdiKDE3LDE3LDE3KTtiYWNrZ3JvdW5kLWNvbG9yOiByZ2IoMjU1LDI1NSwyNTUpO2ZvbnQtc2l6ZTogMTNweDsiPjxzdHJvbmc+RGV0ZXJtaW5pbmcgd2F0dGFnZSByZXF1aXJlbWVudHMuPC9zdHJvbmc+IFdoZW4gYnV5aW5nIGEgZ2VuZXJhdG9yLCBpdCBpcyBpbXBvcnRhbnQgdG8ga25vdyB0aGUgdG90YWwgc3RhcnRpbmcgd2F0dHMgb2YgYWxsIHRoZSBkZXZpY2VzIHlvdSBtYXkgbmVlZCB0byBwb3dlci4gWW91IGNhbiBmaW5kIG91dCB0aGUgdG90YWwgc3RhcnRpbmcgd2F0dHMgYnkgYWRkaW5nIHVwIHRoZSBydW5uaW5nIHdhdHRzIGFuZCBhZGRpdGlvbmFsIHN0YXJ0aW5nIHdhdHRzIG9mIGV2ZXJ5IGRldmljZSB0aGF0IHlvdSBpbnRlbmQgdG8gdXNlLiBUYXAgdGhlIGJ1dHRvbiBiZWxvdyBmb3IgYSBoYW5keSBkb2N1bWVudCB0byBhY2NvbXBsaXNoIHRoaXMuPC9zcGFuPjwvbGk+CjxsaT48c3BhbiBzdHlsZT0iY29sb3I6IHJnYigxNywxNywxNyk7YmFja2dyb3VuZC1jb2xvcjogcmdiKDI1NSwyNTUsMjU1KTtmb250LXNpemU6IDEzcHg7Ij48c3Ryb25nPkhvbWUgc3RhbmRieSBnZW5lcmF0b3JzLjwvc3Ryb25nPiBXaGlsZSBwb3J0YWJsZSBnZW5lcmF0b3JzIGFyZSBncmVhdCBmb3Igc2hvcnQtdGVybSB1c2UsIHlvdSBtaWdodCB3YW50IHRvIGNvbnNpZGVyIGEgaG9tZS1zdGFuZGJ5IGdlbmVyYXRvciBpZiBwb3dlciBvdXRhZ2VzIGFyZSBhIGNvbW1vbiBjb25jZXJuIGluIHlvdXIgYXJlYS4gSG9tZSBzdGFuZGJ5IGdlbmVyYXRvcnMgYXJlIG1vcmUgcmVsaWFibGUsIGVzcGVjaWFsbHkgZm9yIGxvbmctdGVybSB1c2UgYW5kIGFyZSBwb3dlcmZ1bCBlbm91Z2ggdG8gcG93ZXIgeW91ciBlbnRpcmUgaG91c2UuoFdoaWxlIGhvbWUtc3RhbmRieSBnZW5lcmF0b3JzIHJlcXVpcmUgcHJvZmVzc2lvbmFsIGluc3RhbGxhdGlvbiwgaWYgeW91IGxpdmUgaW4gYW4gYXJlYSB0aGF0IGlzIHByb25lIHRvIHBvd2VyIG91dGFnZXMsIHRoZXkgY2FuIGJlIGEgd29ydGh3aGlsZSBpbnZlc3RtZW50Ljwvc3Bhbj4mbmJzcDs8L2xpPgo8L3VsPgo=
```

This is the decoded HTML code (decoded using [https://www.base64decode.org](https://www.base64decode.org)):
```<p style="text-align:left;"><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;">Whatever the season, power outages are a fact of life, which is why having a backup generator is a great way to keep you powered up and your appliances running no matter what happens. Pick out the right generator for you by keeping these helpful facts in mind.</span></p>
<ul>
    <li style="text-align:left;"><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;"><strong>Generator types.</strong> There are two types of generators conventional and inverter.</span></li>
</ul>
<ol>
    <li style="text-align:left;"><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;">A <strong>conventional generator</strong> is typically louder and larger and runs on gasoline. They are usually good at powering a portion of your home (up to 3 or 4 rooms). Conventional generators are great for refrigerators, electric stove tops, AC units and other quality of life devices.</span></li>
    <li style="text-align:left;"><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;">An <strong>inverter generator</strong> is smaller and can run on gasoline or propane. These are usually good for powering one or two devices. An inverter generator is ideal for powering computers, cellphones, radios, and other devices that are critical during emergencies.</span></li>
</ol>
<ul>
    <li><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;"><strong>Determining wattage requirements.</strong> When buying a generator, it is important to know the total starting watts of all the devices you may need to power. You can find out the total starting watts by adding up the running watts and additional starting watts of every device that you intend to use. Tap the button below for a handy document to accomplish this.</span></li>
    <li><span style="color: rgb(17,17,17);background-color: rgb(255,255,255);font-size: 13px;"><strong>Home standby generators.</strong> While portable generators are great for short-term use, you might want to consider a home-standby generator if power outages are a common concern in your area. Home standby generators are more reliable, especially for long-term use and are powerful enough to power your entire house.While home-standby generators require professional installation, if you live in an area that is prone to power outages, they can be a worthwhile investment.</span>&nbsp;</li>
</ul>
```

The error points me to **decodeUriComponent** function.
**Note**: Other base64 encoded html work fine for me, this is the only instance that I got the error and it happens frequently;
### **Workaround**
Use **Base64.atob(string)**

**Comment by dankogai** [comment]
I have reproduced your symptom with:

* Safari 13
* Chrome 85
* Firefox 80

While it worked fine on

* node 14

So it is a matter of `decodeURIComponent()` which is browser built-in.

**Comment by Sephaq** [comment]
I see, I'm using node 14, and using the lib with a react-native

- react-native 0.63.2

**Comment by dankogai** [comment]
I have fixed it and released version 3.5.0.  

Thank you so much for finding this tricky bug.  `escape()` and `unescape()` are fast since they are built-in but they are also ancient voodoo that are deprecated so it would be fixed anyhow.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape

`decodeURIComponent()` and `encodeURIComponent()` are not deprecated but for the truck to work they must be used with `escape()` and `unescape()`.

**Comment by OliverJAsh** [comment]
I am curious to learn why exactly `escape`/`decodeURIComponent` didn't work.

I am unable to reproduce the described bug, using version 3.4.5 of this library, testing in Chrome 87: https://stackblitz.com/edit/typescript-hnjxxs. What am I missing?
